### ISSUE TITLE
Why don't you compute the gradient rather you do backward ?

### ISSUE THREAD

**Issue by HabibAouani** [agil27/TCAV_PyTorch#1]
Hello agil27;

Thank you for your implementation, this really helpful as I'm coding with pytorch.

I have one question:
- Please, why do you compute backward rather than grad (from autograd) in the function:
`
 def generate_gradients(self, c, layer_name):        activation = self.intermediate_activations[layer_name]        activation.register_hook(self.save_gradient)        logit = self.output[:, c]        logit.backward(torch.ones_like(logit), retain_graph=True)        # gradients = grad(logit, activation, retain_graph=True)[0]        # gradients = gradients.cpu().detach().numpy()        gradients = self.gradients.cpu().detach().numpy()        return gradients<br class="Apple-interchange-newline">`

the grad function is commented with "#"

Thank you very much,
Habib

**Comment by agil27** [comment]
Hello HabibAouani,

Thanks for your mail.

Computing gradients using pytorch function grad() also requires the backward process. They are equivalent.

Yuanbiao

在2021-06-09 ***@***.***写道：

Hello agil27;

Thank you for your implementation, this really helpful as I'm coding with pytorch.

I have one question:

Please, why do you compute backward rather than grad (from autograd) in the function:
def generate_gradients(self, c, layer_name): activation = self.intermediate_activations[layer_name] activation.register_hook(self.save_gradient) logit = self.output[:, c] logit.backward(torch.ones_like(logit), retain_graph=True) # gradients = grad(logit, activation, retain_graph=True)[0] # gradients = gradients.cpu().detach().numpy() gradients = self.gradients.cpu().detach().numpy() return gradients<br class="Apple-interchange-newline">

**Comment by HabibAouani** [comment]
Hello;

Thank you for your response !

But you aren't using grad(output,input, retain_graph), rather you're using Backward from logit.

Your Backward function is computing with respect to all layers, not only one layer. I'm I right ?

In your code, you have commented the grad function, which means that it's not used.

Thank you very much for your help,
Habib

**Comment by agil27** [comment]
Hello,

the execution of grad(output,input, retain_graph) requires a backward process, so they are equivalent. (as far as I am concerned)

Yuanbiao

**Comment by HabibAouani** [comment]
Hello Yuanbiao;

Thank you very much.

Habib
